arangod/Aql/IndexJoin/TwoIndicesMergeJoin.h

```diff
@@ -0,0 +1,425 @@
+  LOG_IDX_TWO_NONU_MERGER << "iterator seeking to " << target[0];
+  std::copy(target.begin(), target.end(), _position.begin());
+  exhausted = !_iter->seek(_position);
+
+  LOG_IDX_TWO_NONU_MERGER << "iterator seeked to " << _position[0]
+                          << " exhausted = " << exhausted;
+}
+
+template<typename SliceType, typename DocIdType, typename KeyCompare>
+void TwoIndicesMergeJoin<SliceType, DocIdType,
+                         KeyCompare>::IndexStreamData::reset() {
+  exhausted = !_iter->reset(_position);
+}
+
+template<typename SliceType, typename DocIdType, typename KeyCompare>
+bool TwoIndicesMergeJoin<SliceType, DocIdType,
+                         KeyCompare>::IndexStreamData::next() {
+  return exhausted = !_iter->next(_position, _docId, _projections);
+}
+
+template<typename SliceType, typename DocIdType, typename KeyCompare>
+TwoIndicesMergeJoin<SliceType, DocIdType, KeyCompare>::TwoIndicesMergeJoin(
+    std::vector<Descriptor> descs, std::size_t numKeyComponents) {
+  TRI_ASSERT(descs.size() == FIXED_NON_UNIQUE_INDEX_SIZE_VAR);
+  indexes.reserve(FIXED_NON_UNIQUE_INDEX_SIZE_VAR);
+  currentKeySet.resize(numKeyComponents);
+
+  std::size_t bufferSize = 0;
+  for (auto const& desc : descs) {
+    bufferSize += desc.numProjections + numKeyComponents;
+  }
+
+  sliceBuffer.resize(bufferSize);
+
+  auto keySliceIter = sliceBuffer.begin();
+  auto projectionsIter = sliceBuffer.begin() + numKeyComponents * descs.size();
+  projectionsSpan = {projectionsIter, sliceBuffer.end()};
+  auto docIdIter = documentIds.begin();
+  for (auto& desc : descs) {
+    auto projections = projectionsIter;
+    projectionsIter += desc.numProjections;
+    auto keyBuffer = keySliceIter;
+    keySliceIter += numKeyComponents;
+    auto& idx = indexes.emplace_back(
+        std::move(desc.iter), std::span<SliceType>{keyBuffer, keySliceIter},
+        std::span<SliceType>{projections, projectionsIter}, *(docIdIter++));
+
+    if (maxIter == nullptr ||
+        IndexStreamCompare{}.cmp(*maxIter, idx) == std::weak_ordering::less) {
+      if (maxIter != nullptr) {
+        minIter = maxIter;
+      }
+      maxIter = &idx;
+    } else {
+      minIter = &idx;
+    }
+  }
+
+  TRI_ASSERT(minIter != nullptr);
+  TRI_ASSERT(maxIter != nullptr);
+  TRI_ASSERT(minIter != maxIter);
+  auto isAligned = IndexStreamCompare{}.cmp(*maxIter, *minIter) ==
+                   std::weak_ordering::equivalent;
+  if (!maxIter->exhausted && isAligned) {
+    fillInitialMatch();
+    positionAligned = true;
+  }
+}
+
+template<typename SliceType, typename DocIdType, typename KeyCompare>
+std::pair<bool, std::size_t>
+TwoIndicesMergeJoin<SliceType, DocIdType, KeyCompare>::next(
+    const std::function<bool(std::span<DocIdType>, std::span<SliceType>)>& cb) {
+  size_t amountOfSeeks = 0;
+  while (true) {
+    if (positionAligned) {
+      auto result = produceCrossProduct(cb);
+      if (result.seeks > 0) {
+        amountOfSeeks += result.seeks;
+      }
```

Comment:
is this if actually worth it? you're trading a correctly predicted add for a potentially mispredicted jump